airflow/executors/base_executor.py

```diff
@@ -141,13 +141,27 @@ def fail(self, key):
     def success(self, key):
         self.change_state(key, State.SUCCESS)
 
-    def get_event_buffer(self):
+    def get_event_buffer(self, dag_ids=None):
         """
-        Returns and flush the event buffer
+        Returns and flush the event buffer. In case dag_ids is specified
+        it will only return and flush events for the given dag_ids. Otherwise
+        it returns and flushes all
+
+        :param dag_ids: to dag_ids to return events for, if None returns all
+        :return: a dict of events
         """
-        d = self.event_buffer
-        self.event_buffer = {}
-        return d
+        if not dag_ids:
+            d = self.event_buffer
+            self.event_buffer = {}
+            return d
+        else:
+            d = {}
+            for key, state in list(self.event_buffer.items()):
```

Comment:
why cast to list here? Is this some python 3 thing?

Reply:
Yes, but more importantly I am changing event_buffer in this loop, copying it into a list makes sure not to skip any events.